luarules/gadgets/unit_collision_damage_behavior.lua

```diff
@@ -18,52 +18,67 @@ local fallDamageMagnificationFactor = 14 --the multiplier by which default engin
 local groundCollisionDefID = Game.envDamageTypes.GroundCollision
 local objectCollisionDefID = Game.envDamageTypes.ObjectCollision
 local objectCollisionVelocityThreshold = 3.3 --this defines how fast a unit has to be moving in order to take object collision damage
-local maxImpulseThreshold = 0.08 --incease this to make units move less from impulse. This defines the max impulse damage a unit can take expressed in a percentage multiplier
-local cooldownsFrameThreshold = Game.gameSpeed --this is how long in frames before a unit's maximum impulse limit resets.
+local maxImpulseProportion = 0.04 --incease this to make units move less from impulse. This defines the max impulse damage allowed a unit can take relative to its mass.
 
-local spGetUnitVelocity = Spring.GetUnitVelocity
 local spGetUnitHealth = Spring.GetUnitHealth
-local mathmin = math.min
+local spGetUnitVelocity = Spring.GetUnitVelocity
+local spSetUnitVelocity = Spring.SetUnitVelocity
+local mathMin = math.min
+local mathMax = math.max
+local mathCeil = math.ceil
 
 local fallDamageMultipliers = {}
 local unitsMaxImpulse = {}
 local weaponDefIDImpulses = {}
-local overImpulseCooldowns = {}
-local transportingUnits = {}
+local transportedUnits = {}
 local unitMasses = {}
-
-local currentModulusFrame = 0
+local weaponDefIgnored = {}
+local unitInertiaCheckFlags = {} -- track inertia via velocity (units have constant mass)
+local gameFrame = 0
+local velocityWatchDuration = 3
 
 for unitDefID, unitDef in ipairs(UnitDefs) do
 	local fallDamageMultiplier = unitDef.customParams.fall_damage_multiplier or 1.0
-	fallDamageMultipliers[unitDefID] = fallDamageMultiplier*fallDamageMagnificationFactor
-	unitsMaxImpulse[unitDefID] = unitDef.mass/maxImpulseThreshold
+	fallDamageMultipliers[unitDefID] = fallDamageMultiplier * fallDamageMagnificationFactor
+	unitsMaxImpulse[unitDefID] = unitDef.mass / maxImpulseProportion
 	unitMasses[unitDefID] = unitDef.mass
 end
 
+for name, weaponDefID in pairs(Game.envDamageTypes or {}) do
```

Comment:
In what circumstances is `Game.envDamageTypes` nil?

Reply:
yep, fixed

Reply:
```suggestion
for name, weaponDefID in pairs(Game.envDamageTypes) do
```
I'm saying why is the nil coalescing to an empty table needed, when `Game.envDamageTypes` can never be nil